### PR TITLE
Add docstring to `BaseStorage` method interfaces.

### PR DESCRIPTION
**Pull request by ytsmiling**
## Motivation
Storage class implementations for each backend handles error in a different manner. Additionally, undocumented return values force users to read implementations of each method. This PR addresses these issues as a step in #1170. Any inconsistent behaviors with this docstring will be fixed in subsequent PRs.

## Description of the changes
Add docstrings to methods in `BaseStorage` class.

### CODE REVIEW

**Comment by ytsmiling** [comment]
Current doc and implementation conflicts with #1174 and allows to change the attributes of WAITING trials, which should be fixed.

File: optuna/storages/base.py [hunk 1]
```diff
@@ -29,157 +28,539 @@ class BaseStorage(object, metaclass=abc.ABCMeta):
...
+        """
 
         raise NotImplementedError
 
     @abc.abstractmethod
-    def set_trial_state(self, trial_id, state):
-        # type: (int, TrialState) -> bool
+    def set_trial_state(self, trial_id: int, state: TrialState) -> bool:
+
+        """Update a state of a specified trial.
```
Comment (ytsmiling) [inline optuna/storages/base.py:284]: I don't think we should allow the "RUNNING → WAITING" state change. (Both this doc and current implementation allows it now.)

Comment (ytsmiling) [reply optuna/storages/base.py:261]: This PR won't address this issue.

File: optuna/storages/base.py [hunk 2]
```diff
@@ -194,31 +529,79 @@ def get_best_trial(self, study_id):
...
+            :exc:`RuntimeError`:
+                If the trial is already finished.
+        """
         if trial_state.is_finished():
             trial = self.get_trial(trial_id)
             raise RuntimeError(
                 "Trial#{} has already finished and can not be updated.".format(trial.number)
             )
+
+    def load(self, study_id: int) -> None:
```
Comment (sile) [inline optuna/storages/base.py:605]: Do we need to introduce this new method in this PR?
I think that the responsibility of this PR should be simply adding the doc to **the existing methods**. So, IMO, it feels better to separate this part into another PR (and it could be easier to review if the implementation and usage code of this method are also included in the PR).

Comment (ytsmiling) [reply optuna/storages/base.py:605]: I left this method because an internal committee discussion reached an agreement that we can introduce the method here, but my opinion is neutral.
I generally agree the opinion and also don't think it's a problem to change the prior decision.
I'll remove the method later.

Comment (sile) [reply optuna/storages/base.py:605]: > an internal committer discussion reached an agreement that we can introduce the method here.
> ...
> I'll remove the method later.

Oh, sorry for my comment disregarding the previous decision, but I would appreciate it if you could do it.

Comment (ytsmiling) [reply optuna/storages/base.py]: Removed the method in https://github.com/optuna/optuna/pull/1175/commits/cfd6f2128ced8a5644b66ce4040133b0e460b687.

Comment (sile) [reply optuna/storages/base.py]: Thanks!

File: optuna/storages/base.py [hunk 3]
```diff
@@ -29,158 +28,494 @@ class BaseStorage(object, metaclass=abc.ABCMeta):
...
+
+        Args:
+            trial_id:
+                Trial ID of a trial to update its state.
+            state:
+                New state of the trial.
+
+        Returns:
+            :py:obj:`True` if the state is successfully updated.
+            :py:obj:`False` if the state is kept the same.
```
Comment (hvy) [inline optuna/storages/base.py:273]: A detail but I don't think you need the `:py` part?

File: optuna/storages/base.py [hunk 4]
```diff
@@ -29,158 +28,494 @@ class BaseStorage(object, metaclass=abc.ABCMeta):
...
         raise NotImplementedError
 
     @abc.abstractmethod
-    def get_all_study_summaries(self):
-        # type: () -> List[study.StudySummary]
+    def get_all_study_summaries(self) -> List[study.StudySummary]:
+        """Read a list of `study.StudySummary` objects.
+
+        Returns:
+            A list of `study.StudySummary` objects.
```
Comment (hvy) [inline optuna/storages/base.py:228]: Why not make it a proper reference?

File: optuna/storages/base.py [hunk 5]
```diff
@@ -29,158 +28,494 @@ class BaseStorage(object, metaclass=abc.ABCMeta):
...
+            trial_id:
+                Trial ID of a trial to update its state.
+            state:
+                New state of the trial.
+
+        Returns:
+            :py:obj:`True` if the state is successfully updated.
+            :py:obj:`False` if the state is kept the same.
+            The latter happens when this method tries to update the state of
+            RUNNING trial to RUNNING.
```
Comment (hvy) [inline optuna/storages/base.py:275]: Let's make those proper references as well.

File: optuna/storages/base.py [hunk 6]
```diff
@@ -29,158 +28,494 @@ class BaseStorage(object, metaclass=abc.ABCMeta):
...
-    def get_n_trials(self, study_id, state=None):
-        # type: (int, Optional[TrialState]) -> int
+    def get_n_trials(self, study_id: int, state: Optional[TrialState] = None) -> int:
+        """Count the number of trials in a specified study.
+
+        Args:
+            study_id:
+                Study ID of a study to count trials.
+            state:
+
```
Comment (hvy) [inline optuna/storages/base.py:492]: Ops, missing explanation?

File: optuna/storages/base.py [hunk 7]
```diff
@@ -29,158 +28,494 @@ class BaseStorage(object, metaclass=abc.ABCMeta):
...
-    def delete_study(self, study_id):
-        # type: (int) -> None
+        Args:
+            study_name:
+                Name of a new study to create.
 
+        Returns:
+            Study ID, which is an unique id among studies, of the created study.
+
+        Raises:
```
[the remark this answers is not in the document]
Comment (hvy) [reply optuna/storages/base.py:44]: We (or at least I) haven't really paid much attention to documenting exceptions, but we definitely should.

**Comment by ytsmiling** [comment]
@hvy Thank you for your review. I appreciate your follow-up PR suggestion, and I'd like to see it.

File: optuna/storages/base.py [hunk 8]
```diff
@@ -29,158 +28,494 @@ class BaseStorage(object, metaclass=abc.ABCMeta):
...
+            :exc:`optuna.exceptions.DuplicatedStudyError`:
+                If a study with the same `study_name` already exists.
```
[the remark this answers is not in the document]
Comment (ytsmiling) [reply optuna/storages/base.py]: Thank you. I fixed all of them I could found.

File: optuna/storages/base.py [hunk 9]
```diff
@@ -29,158 +28,494 @@ class BaseStorage(object, metaclass=abc.ABCMeta):
...
+            study_id:
+                Study ID of a study to update.
+            direction:
+                Either StudyDirection.MAXIMIZE or StudyDirection.MINIMIZE.
+
+        Raises:
+            :exc:`KeyError`:
+                If no study with the matching `study_id` exists.
+            :exc:`ValueError`:
+                If `direction` is already set and the passed `direction` conflicts with it.
```
[the remark this answers is not in the document]
Comment (ytsmiling) [reply optuna/storages/base.py]: Thank you. I made the point explicit.

File: optuna/storages/base.py [hunk 10]
```diff
@@ -29,158 +28,494 @@ class BaseStorage(object, metaclass=abc.ABCMeta):
...
+        """
+        raise NotImplementedError
 
+    @abc.abstractmethod
+    def set_study_system_attr(self, study_id: int, key: str, value: Any) -> None:
+        """Register an ontuna-internal attribute to a study.
```
[the remark this answers is not in the document]
Comment (ytsmiling) [reply optuna/storages/base.py]: Thank you. Nich cache.

File: optuna/storages/base.py [hunk 11]
```diff
@@ -29,158 +28,494 @@ class BaseStorage(object, metaclass=abc.ABCMeta):
...
-    def set_study_system_attr(self, study_id, key, value):
-        # type: (int, str, Any) -> None
+    def set_study_direction(self, study_id: int, direction: study.StudyDirection) -> None:
+        """Register a direction of optimization problem to a study.
+
+        Args:
+            study_id:
+                Study ID of a study to update.
+            direction:
+                Either StudyDirection.MAXIMIZE or StudyDirection.MINIMIZE.
```
[the remark this answers is not in the document]
Comment (ytsmiling) [reply optuna/storages/base.py]: Thank you. I fixed the references.

File: optuna/storages/base.py (hunk 3 above)
Comment (ytsmiling) [reply optuna/storages/base.py]: Removed.

File: optuna/storages/base.py (hunk 4 above)
Comment (ytsmiling) [reply optuna/storages/base.py]: Thank you. I fixed them.

File: optuna/storages/base.py (hunk 5 above)
Comment (ytsmiling) [reply optuna/storages/base.py]: Thank you. I added references.

File: optuna/storages/base.py (hunk 6 above)
Comment (ytsmiling) [reply optuna/storages/base.py:495]: Thank you. I added the missing doc.

File: optuna/storages/base.py [hunk 12]
```diff
@@ -29,158 +28,494 @@ class BaseStorage(object, metaclass=abc.ABCMeta):
...
+        Raises:
+            :exc:`KeyError`:
+                If no study with the matching `study_id` exists.
+            :exc:`ValueError`:
+                If `direction` is already set and the passed `direction` conflicts with it.
+        """
+        raise NotImplementedError
 
+    @abc.abstractmethod
+    def set_study_system_attr(self, study_id: int, key: str, value: Any) -> None:
```
[the remark this answers is not in the document]
Comment (ytsmiling) [reply optuna/storages/base.py]: I agree with that and changed the order.

File: optuna/storages/base.py [hunk 13]
```diff
@@ -29,158 +28,497 @@ class BaseStorage(object, metaclass=abc.ABCMeta):
...
-        # type: (int, Optional[TrialState]) -> int
+    def get_all_trials(self, study_id: int, deepcopy: bool = True) -> List["FrozenTrial"]:
+        """Read all trials in a specified study.
+
+        Args:
+            study_id:
+                Study ID of a study to read trials from.
+            deepcopy:
+                Whether copy the list of trials before returning.
+                Set :py:obj:`True` when you might update the list or elements of the list.
```
Comment (hvy) [inline optuna/storages/base.py:475]:
```suggestion
                Set :obj:`True` when you might update the list or elements of the list.
```

File: optuna/storages/base.py [hunk 14]
```diff
@@ -29,158 +28,497 @@ class BaseStorage(object, metaclass=abc.ABCMeta):
...
+        Args:
+            trial_id:
+                Trial ID of a trial to set intermediate value.
+            step:
+                Step of the trial (e.g., Epoch of neural network training).
+            intermediate_value:
+                Reported value within an evaluation of an objective function.
+
+        Returns:
+            Return :py:obj:`False` when the intermediate of the step already exists.
```
Comment (hvy) [inline optuna/storages/base.py:393]:
```suggestion
            Return :obj:`False` when the intermediate of the step already exists.
```

File: optuna/storages/base.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (hvy) [reply optuna/storages/base.py]: Thanks!

File: optuna/storages/base.py [hunk 15]
```diff
@@ -29,158 +28,497 @@ class BaseStorage(object, metaclass=abc.ABCMeta):
...
+                Study ID of a study to update.
+            direction:
+                Either :obj:`~optuna.study.StudyDirection.MAXIMIZE` or
+                :obj:`~optuna.study.StudyDirection.MINIMIZE`.
+
+        Raises:
+            :exc:`KeyError`:
+                If no study with the matching ``study_id`` exists.
+            :exc:`ValueError`:
+                If `direction` is already set and the passed ``direction`` is the opposite
```
Comment (hvy) [inline optuna/storages/base.py:120]: Might be missing something but should this not be plain text?
```suggestion
                If direction is already set and the passed ``direction`` is the opposite
```

File: optuna/storages/base.py [hunk 16]
```diff
@@ -194,29 +532,73 @@ def get_best_trial(self, study_id):
...
+        return self.get_trial(trial_id).system_attrs
+
+    def remove_session(self) -> None:
+        """Clean up all connections to a database."""
         pass
 
-    def check_trial_is_updatable(self, trial_id, trial_state):
-        # type: (int, TrialState) -> None
+    def check_trial_is_updatable(self, trial_id: int, trial_state: TrialState) -> None:
+        """Check whether a `trial_state` is updatable.
```
Comment (hvy) [inline optuna/storages/base.py:589]: How about two backticks, or as plain text
```suggestion
        """Check whether a trial state is updatable.
```
